mqtt/mqtt-bridge/src/messages.rs

```diff
@@ -81,20 +83,53 @@ impl<S> StoreMqttEventHandler<S> {
-                .filter(|stripped_topic| mapper.topic_filter.matches(stripped_topic))
-                .map(|stripped_topic| match mapper.topic_settings.out_prefix() {
-                    Some(remote_prefix) => format!("{}/{}", remote_prefix, stripped_topic),
-                    None => stripped_topic.to_string(),
-                })
+            if mapper.topic_filter.matches(topic_name) {
+                mapper
+                    .topic_settings
+                    .in_prefix()
+                    // maps if local does not have a value it uses the topic that was received,
+                    // else it checks that the received topic starts with local prefix and removes the local prefix
+                    .map_or(Some(topic_name), |in_prefix| {
+                        topic_name.strip_prefix::<&str>(in_prefix)
+                    })
+                    .map(|stripped_topic| match mapper.topic_settings.out_prefix() {
+                        Some(out_prefix) => {
+                            if mapper.topic_settings.in_prefix().is_none() {
+                                // inPrefix is empty so it adds the topic separator between remote prefix
+                                // and topic
+                                format!("{}{}{}", out_prefix, TOPIC_SEPARATOR, stripped_topic)
+                            } else {
+                                format!("{}{}", out_prefix, stripped_topic)
+                            }
+                        }
+                        None => {
+                            if mapper.topic_settings.in_prefix().is_some() {
+                                stripped_topic
+                                    .strip_prefix(TOPIC_SEPARATOR)
+                                    .unwrap_or(stripped_topic)
+                                    .to_string()
+                            } else {
+                                // in case there was no separator because inPrefix is empty
+                                stripped_topic.to_string()
+                            }
+                        }
+                    })
+                    .and_then(|transformed_topic| {
+                        // transform_topic can be empty when topic is # and outPrefix is empty and it matches on inPrefix
+                        // example topic: #, inPrefix: local/messages, outPrefix: "" and message is sent with topic local/messages
+                        if transformed_topic.is_empty() {
+                            warn!(
+                                "topic {} was matched with {:#?}, but remote topic is not valid",
+                                topic_name, mapper.topic_settings
+                            );
+                            None
+                        } else {
+                            Some(transformed_topic)
+                        }
+                    })
+            } else {
+                None
```

Comment:
Code is rather complicated here we should probably refactor it. Maybe split in multiple functions will help.

Reply:
After Varun comments and discussion I changed the code so it doesn't add / by default. This made the code simple and also less cases to explain in doc for customers.